Lms/add teacher notification email frequency

anathomical opened the pull request:
## WHAT
- Add a column to `TeacherInfo` to store desired email notification frequency
- Add `after_create` hook on `User` to set a default value for this column on newly-created users
- Tweak the `TeacherInfosController` so that calls to the `create` action will use `find_or_create` so that when we create `TeacherInfo` records early in our `after_create` hook we don't break later calls to the `create` action by running into uniqueness constraints
## WHY
We want users to be able to specify how frequently the receive rollup emails to let them strike their own balance between number of emails and freshness of data.
## HOW
- Add a column to the TeacherInfo table
- Add a `User` `after_create` hook to set default values
- Update `TeacherInfosController` to accept the new value so that we can use it to update this value moving forward

### Notion Card Links
https://www.notion.so/quill/Send-a-roll-up-email-of-student-related-events-to-teachers-at-a-frequency-of-their-choosing-a074aabc41eb41be84f732db7a63aa49?pvs=4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Deploying now
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Add a new TeacherInfo field for storing email notification frequency

Also update the TeacherInfosController `create` action so that it won't create additional records if one already exists for a given user.

anathomical committed:
Set up code to set default values for notification_frequency on new users

anathomical committed:
Remove two migrations from structure from a different branch

anathomical committed:
Remove more migrations from other branches from structure.sql

anathomical committed:
Fix db/structure, hopefully

anathomical committed:
Clean up specs to work with new more common TeacherInfo records

anathomical committed:
Tweak ClassroomManager teacher info modal to appearance logic

We used to be able to depend on the simple presence/absence of teacher_info records for the teacher to determine whether the teacher should see the modal asking for their grade levels and subject areas.  Now that we're storing teacher_info for additional info not related to this modal (email frequency), we need a slightly more precise thing to check to figure out whether the user needs to see the modal.  I've settled on minimum_grade_level for now.

anathomical committed:
Update more specs using teacher_info

brendanshean commented on services/QuillLMS/app/controllers/teacher_infos_controller.rb:
```diff
@@ -2,42 +2,53 @@
     }, status: 200
   end
 
+  private def set_teacher_info
```
The logic here is a little surprising given the conventions for `create` and `update` in Rails controllers.

Unless there's a good reason not to, I think create should be reserved the case where `teacher_info` is created only.  Then use update when the object already exists;.

brendanshean commented on services/QuillLMS/app/controllers/teacher_infos_controller.rb:
```diff
@@ -2,42 +2,53 @@
-      teacher_info.teacher_info_subject_areas.destroy_all
-      subject_area_ids.each { |subject_area_id| TeacherInfoSubjectArea.create!(teacher_info: teacher_info, subject_area_id: subject_area_id) }
+      @teacher_info.teacher_info_subject_areas.destroy_all
+      subject_area_ids.each { |subject_area_id| TeacherInfoSubjectArea.create!(teacher_info: @teacher_info, subject_area_id: subject_area_id) }
```
I like using the create directly from the base model:

`subject_area_ids.each { |subject_area_id| @teacher_info.teacher_info_subject_areas(subject_area_id: subject_area_id) }`

This pattern highlights the has_many relationship but also makes the context of the creation the @teacher_info object which is consistent with the controller.

brendanshean commented on services/QuillLMS/app/models/user.rb:
```diff
@@ -860,5 +861,9 @@ def school_or_district_premium?
   end
+
+  private def generate_default_notification_email_frequency
+    TeacherInfo.create!(user: self, notification_email_frequency: TeacherInfo::DAILY_EMAIL)
```
I would use:
`teacher_info.create!(notification_email_frequency: TeacherInfo::DAILY_EMAIL)`

since it highlights the has_one relationship with teacher_info and has the implicit user context.

brendanshean commented on services/QuillLMS/spec/models/segment_integration/user_spec.rb:
```diff
@@ -101,6 +101,8 @@
     let(:teacher_info) { create(:teacher_info, user: teacher) }
```
What about using the teacher_info from the teacher:

`let(:teacher_info) { teacher.teacher_info }`

then you could get rid of lines 104-105

brendanshean reviewed (approved):
There's a comment about create/update in the controller, but nothing that requires a re-review.  Nice work on this.

anathomical committed:
Make tweaks based on code review

anathomical committed:
Use update action for teacher info when signing up

Since we're already creating a teacher_info record for new users in a hook, so making a create call here will error

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/add-teacher-notification-email-frequency
